### ISSUE TITLE
[ML] Nudge the model selection to prefer smaller models with similar results

### ISSUE THREAD

**Issue by valeriy42** [elastic/ml-cpp#1481]
We should prefer smaller models if the loss function values are about the same. This will result in smaller memory requirements, faster inference and better reproducibility of the training runs.

* It seems that the stopping criterion is activated to late in the process s.t. we already overfitting for quite a while. We need to verify that the stopping criterion works as we want it and adjust it if necessary.
* In `CBoostedTreeImpl::captureBestHyperparameters` we should include the forest size (or better the total number of nodes and leaves in the forest) as the preference if the loss is similar (e.g. within 1%).

**Comment by valeriy42** [comment]
Relative comparison of losses and forest sizes:

```c++
#include <boost/math/special_functions/relative_difference.hpp> 

//...
void CBoostedTreeImpl::captureBestHyperparameters(const TMeanVarAccumulator& lossMoments,
                                                  std::size_t maximumNumberTrees) {
{
    double loss{lossAtNSigma(1.0, lossMoments)};
    if (boost::math::relative_difference(loss, m_BestForestTestLoss) < 0.01 && maximumNumberTrees < m_BestForestTestLoss.maximumNumberTrees()) {
        m_BestForestTestLoss = loss;
        m_BestHyperparameters = CBoostedTreeHyperparameters{
            m_Regularization,       m_DownsampleFactor, m_Eta,
            m_EtaGrowthRatePerTree, maximumNumberTrees, m_FeatureBagFraction};
    }
}
```

**Comment by valeriy42** [comment]
@tveasey did I miss something?

**Comment by tveasey** [comment]
I think this is good except I would count nodes rather than just trees.

My inclination would also be to roll this into the loss values supplied to Bayesian Optimisation because it should also adjust (stabilise) how we search the parameters. This can use the same small scaling idea, but will be more fiddly and may need some experimentation.

So for something very simple we could put the number of trees in the forest on the scale of 0.01 * mean(loss surface). So for example, in `CBoostedTreeImpl::selectNextHyperparameters` something like:
1. Keep track of the "average forest number of nodes" .
2. Add `0.01 * "forest number nodes" * E[GP] / "average forest number nodes"` to `meanLoss`.

The one thing is we're effectively changing the constant of proportionality slightly over the rounds because of changes to "average forest number nodes". We seed with random search, so we could use these iterations to get a stable estimate of "average forest number nodes", initialise BO only at this point (with all rounds parameters losses up to that point) and fix it from then on. Alternatively it may just not matter. This is after all only about nudging us to more efficient solutions when the difference predictive accuracy is small.

**Comment by valeriy42** [comment]
Initial try results:
![image](https://user-images.githubusercontent.com/1292899/94535376-cb304480-0241-11eb-8773-50529f0024df.png)
